botorch/models/utils/inducing_point_allocators.py

```diff
@@ -0,0 +1,388 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+r"""
+References
+
+.. [burt2020svgp]
+    David R. Burt and Carl Edward Rasmussen and Mark van der Wilk,
+    Convergence of Sparse Variational Inference in Gaussian Process Regression,
+    Journal of Machine Learning Research, 2020,
+    http://jmlr.org/papers/v21/19-1015.html.
+
+.. [chen2018dpp]
+    Laming Chen and Guoxin Zhang and Hanning Zhou, Fast greedy MAP inference
+    for determinantal point process to improve recommendation diversity,
+    Proceedings of the 32nd International Conference on Neural Information
+    Processing Systems, 2018, https://arxiv.org/abs/1709.05135.
+
+.. [moss2023ipa]
+    Henry B. Moss and Sebastian W. Ober and Victor Picheny,
+    Inducing Point Allocation for Sparse Gaussian Processes
+    in High-Throughput Bayesian Optimization,Proceedings of
+    the 25th International Conference on Artificial Intelligence
+    and Statistics, 2023, https://arxiv.org/pdf/2301.10123.pdf.
+
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Callable, Union
+
+import torch
+from botorch.models.model import Model
+
+from botorch.utils.probability.utils import ndtr as Phi, phi
+from gpytorch.module import Module
+from linear_operator.operators import LinearOperator
+from torch import Tensor
+
+NEG_INF = -(torch.tensor(float("inf")))
```

Comment:
Why is this?

Reply:
A `float` just requires 32-64 bits in stack-allocated memory, so even `torch.tensor(-float("inf"))` would 1. store the float corresponding to `inf`, 2. negate it in place, since the variable is not used anywhere else, 3. allocate a `Tensor` object, which is generally heap-allocated and keeps track of a lot more information than a `float`: `requires_grad`, `grad`, `device`, `dtype`, etc.

`-(torch.tensor(float("inf")))` on the other hand, allocates two `Tensor` objects: First by calling `torch.tensor(float("inf"))` and subsequently by negating the first `Tensor`, which allocates a second `Tensor` object with all the extraneous variables. Clearly this would never be a bottleneck here, but in similar places with larger tensors, this could halve the memory footprint.

Reply:
got it! Thanks for explaining